modules/common/src/main/java/StringUtils.java

```diff
@@ -1,62 +1,77 @@
-
-        sb.append('"');
-        for (char c : sArr) {
-            boolean safe = c >= ' ' && c <= '~';
-            if (safe) switch(c) {
+        int safeIdx = 0;
+        for (int i = 0; i < len; i++) {
+            char c = sArr[i];
+            if (c >= ' ' && c <= '~') switch(c) {
                 case '<': case '>': case '&': case '"':
                 case '\'': case '\\': case '`':
-                  safe = false;
+                    break;  // cur char is bad, escape it
+                default:
+                    continue;  // char is OK, continue scan.
             }
-
-            if (safe) {
-                sb.append(c);
-            } else {
-                sb.append(c > '\u00ff' ?
-                    c > '\u0fff' ? "\\u" : "\\u0" :
-                    c > '\u000f' ? "\\u00" : "\\u000");
-                sb.append(Integer.toHexString(c));
+            // this code runs when char is either out of alphanumeric range OR
+            // char is restricted.
+            if (sb == null) {
+              sb = new StringBuilder(c <= '~' ? len + 22 : len * 6 + 2);
+              sb.append('"');
             }
+            sb.append(s, safeIdx, i);
+            sb.append(new char[] { '\\', 'u',
+                DIGITS[c >>> 12],
+                DIGITS[(c >>> 8) & 0xf],
+                DIGITS[(c >>> 4) & 0xf],
+                DIGITS[c & 0xf]
+            });
+            safeIdx = i + 1;
         }
+        if (sb == null) return "\"" + s + "\"";
+        sb.append(s, safeIdx, len);
         sb.append('"');
-
         return sb.toString();
     }
 
     public static String escapeHtml(String s) {
-        // Extensively profiled with jmh. It is faster to pass
-        // around the String reference than to reuse the char
-        // array.
-        for (char c : s.toCharArray()) {
-            switch (c) {
+        char[] sArr = s.toCharArray();
+        for (int i = 0, end = sArr.length; i < end; i++) {
+            switch (sArr[i]) {
                 case '<': case '>': case '&': case '"': case '\'':
-                  return reallyEscapeHtml(s);
+                  StringBuilder sb = new StringBuilder(end + 10);
+                  sb.append(s, 0, i);
+                  escapeHtml(sb, sArr, i, end);
+                  return sb.toString();
             }
         }
         return s;
     }
 
-    private static String reallyEscapeHtml(String s) {
-        char[] sArr = s.toCharArray();
-        StringBuilder sb = new StringBuilder(sArr.length + 10);
-        for (char c : sArr) {
-            switch (c) {
-                case '<': sb.append("&lt;"); continue;
-                case '>': sb.append("&gt;"); continue;
-                case '&': sb.append("&amp;"); continue;
-                case '"': sb.append("&quot;"); continue;
-                case '\'': sb.append("&#39;"); continue;
-                default: sb.append(c);
+    public static void escapeHtml(StringBuilder sb, char[] sArr, int start, int end) {
```

Comment:
can this be private?

Reply:
I'm using it directly in the url regex followup

Reply:
hence the odd api